File: withu/Networking/GenerationQuota.swift
//
//  GenerationQuota.swift
//  withu (iOS)
//
//  AI 생성 횟수 관리.
//    1) 무료 하루 한도 (freeDailyLimit)        — 0 (무료 없음, 캔디로만 생성)
//    2) 구독자 하루 한도 (subscriberDailyLimit) — 구독 활성 시, 매일 리셋
//    3) 충전 크레딧 = 캔디 (credits)            — 캔디 팩 구매분, 만료 없음
//
//  소비 우선순위: 오늘의 (구독) 한도 → 그 다음 캔디.
//
//  ⚠️ 클라이언트 측 저장(App Group UserDefaults). 진짜 비용 방어는 서버 rate limit +
//     OpenAI 대시보드 hard cap 이 담당. 신규 키만 추가 — 기존 스키마 불변.
//

import Foundation

enum GenerationQuota {
    /// 무료 사용자 하루 한도. 0 = 무료 없음 — 생성은 캔디(credits)로만.
    static let freeDailyLimit = 0
    /// 구독 사용자 하루 한도. (무제한으로 하려면 아주 큰 값.)
    static let subscriberDailyLimit = 100

    /// StoreManager 가 앱 시작/구매 시 갱신. true 면 구독 한도 적용.
    static var isSubscriber = false

    private static let countKey   = "withu.genQuota.count.v1"
    private static let dateKey    = "withu.genQuota.date.v1"    // yyyymmdd 정수
    private static let creditsKey = "withu.genQuota.credits.v1"
    /// 마지막으로 로컬에 반영한 서버 잔액 기준선 — 서버 잔액 '증가분'만 로컬에 더하기 위함.
    private static let lastSyncedServerCreditsKey = "withu.genQuota.lastSyncedServer.v1"

    private static var defaults: UserDefaults? {
        UserDefaults(suiteName: SharedAppState.groupID)
    }

    /// 오늘 날짜 yyyymmdd. 자정 넘으면 값이 바뀌어 일일 카운트 자동 리셋.
    private static func todayStamp(_ calendar: Calendar = .current) -> Int {
        let c = calendar.dateComponents([.year, .month, .day], from: Date())
        return (c.year ?? 0) * 10000 + (c.month ?? 0) * 100 + (c.day ?? 0)
    }

    /// 오늘의 한도 (구독 여부에 따라).
    static var dailyAllowance: Int {
        isSubscriber ? subscriberDailyLimit : freeDailyLimit
    }

    /// 오늘 일일 한도에서 사용한 횟수.
    static func usedToday() -> Int {
        guard let d = defaults else { return 0 }
        return d.integer(forKey: dateKey) == todayStamp() ? d.integer(forKey: countKey) : 0
    }

    /// 충전 크레딧 잔액 (만료 없음).
    static func credits() -> Int {
        defaults?.integer(forKey: creditsKey) ?? 0
    }

    /// 지금 더 만들 수 있는 총 횟수 = (오늘 남은 일일 한도) + 크레딧.
    static func remainingToday() -> Int {
        #if DEBUG
        return 9999   // 개발 빌드 — 무제한 테스트 (출시 빌드는 실제 한도)
        #else
        return max(0, dailyAllowance - usedToday()) + credits()
        #endif
    }

    /// n 회를 만들 수 있나.
    static func canGenerate(_ n: Int = 1) -> Bool {
        remainingToday() >= n
    }

    /// n 회 사용 기록. 오늘 한도부터 차감, 부족분은 크레딧에서.
    static func record(_ n: Int = 1) {
        #if DEBUG
        return   // 개발 빌드 — 차감 안 함 (무제한 테스트)
        #else
        guard let d = defaults, n > 0 else { return }
        let today = todayStamp()
        let usedBase = d.integer(forKey: dateKey) == today ? d.integer(forKey: countKey) : 0

        let dailyLeft = max(0, dailyAllowance - usedBase)
        let fromDaily = min(n, dailyLeft)
        let fromCredits = n - fromDaily

        d.set(today, forKey: dateKey)
        d.set(usedBase + fromDaily, forKey: countKey)
        if fromCredits > 0 {
            d.set(max(0, credits() - fromCredits), forKey: creditsKey)
        }
        #endif
    }

    /// 퀄리티별 캔디 비용 — 낮음 1 · 보통 2 · 높음 3.
    static func cost(forQuality quality: String) -> Int {
        switch quality {
        case "high":   return 3
        case "medium": return 2
        default:       return 1   // low
        }
    }

    /// 횟수 팩 구매 시 크레딧 적립.
    static func addCredits(_ n: Int) {
        guard let d = defaults, n > 0 else { return }
        d.set(credits() + n, forKey: creditsKey)
    }

    /// 서버 잔액(entitlement.credits)을 로컬에 반영.
    /// 서버는 /generate 로 캔디를 차감하지 않으므로(로컬 권위) 서버 잔액은 '누적 적립 총액'.
    /// 따라서 절대값으로 덮어쓰면 로컬 차감이 재실행/새로고침마다 되돌아온다(캔디 안 닳는 버그).
    /// → 서버 잔액이 '증가한 만큼(새 구매/적립분)'만 로컬에 더한다.
    static func syncCreditsUp(to serverCredits: Int) {
        guard let d = defaults else { return }
        if d.object(forKey: lastSyncedServerCreditsKey) == nil {
            // 최초 동기화: 서버 적립을 한 번 끌어옴(신규 로그인 대비). 이후엔 증가분만.
            if serverCredits > credits() { d.set(serverCredits, forKey: creditsKey) }
            d.set(serverCredits, forKey: lastSyncedServerCreditsKey)
            return
        }
        let lastSynced = d.integer(forKey: lastSyncedServerCreditsKey)
        let delta = serverCredits - lastSynced
        if delta > 0 {
            d.set(credits() + delta, forKey: creditsKey)   // 새로 적립된 만큼만 더함
        }
        d.set(serverCredits, forKey: lastSyncedServerCreditsKey)
    }

    /// 화면 표시용 보유 캔디 = 실제로 쓸 수 있는 로컬 잔액(credits).
    /// canGenerate 가 로컬 credits 만 보므로, 배지도 같은 값을 보여 과대표시(쓸 수 없는데 숫자만 큼)를 막는다.
    /// 서버 적립분은 syncCreditsUp 으로 이미 로컬에 반영됨.
    static func displayedCandy() -> Int {
        credits()
    }

    // MARK: - 테스트 캔디 코드 (TestFlight/샌드박스 전용)

    /// 입력하면 +20 캔디(로컬). 운영 App Store 빌드에서는 동작 안 함 → 무료 캔디 악용 방지.
    static let testCandyCode = "CANDY20"
    static let testCandyAmount = 20

    /// 테스트 캔디 코드 허용 여부. App Store 운영 빌드(receipt)는 false, TestFlight/샌드박스(sandboxReceipt)·DEBUG 만 true.
    static var allowsTestCandyCode: Bool {
        #if DEBUG
        return true
        #else
        return Bundle.main.appStoreReceiptURL?.lastPathComponent == "sandboxReceipt"
        #endif
    }
}

File: android/app/src/main/java/com/seoyoung/withu/quota/GenerationQuota.kt
package com.seoyoung.withu.quota

import com.seoyoung.withu.BuildConfig
import com.seoyoung.withu.shared.SharedAppState
import java.util.Calendar
import kotlin.math.max
import kotlin.math.min

/**
 * AI 생성 횟수 관리 — iOS GenerationQuota.swift 포팅 (스펙 07 §3.2).
 *   1) 무료 하루 한도 freeDailyLimit = 0 (무료 없음 — 캔디로만 생성)
 *   2) 충전 크레딧 = 캔디 (만료 없음)
 *
 * ⚠️ 캔디 잔액의 권위는 로컬 (서버는 /generate 에서 차감하지 않음).
 *    서버 잔액은 syncCreditsUp 으로 '증가분만' 반영 — 절대값 덮어쓰기 금지
 *    (덮어쓰면 로컬 차감이 새로고침마다 되돌아오는 '캔디 안 닳는' 버그).
 * 저장소: 공유 SharedPreferences — 키 문자열은 iOS 원문 유지.
 */
object GenerationQuota {
    /** 무료 사용자 하루 한도. 0 = 무료 없음. */
    const val freeDailyLimit: Int = 0

    // 테스트 캔디 코드 (TestFlight/샌드박스 대응 — Android 는 DEBUG 전용)
    const val testCandyCode: String = "CANDY20"
    const val testCandyAmount: Int = 20

    /** 테스트 캔디 코드 허용 여부 — 운영 릴리즈 무료 캔디 악용 방지. */
    val allowsTestCandyCode: Boolean get() = BuildConfig.DEBUG

    private const val COUNT_KEY = "withu.genQuota.count.v1"
    private const val DATE_KEY = "withu.genQuota.date.v1"        // yyyymmdd 정수
    private const val CREDITS_KEY = "withu.genQuota.credits.v1"
    /** 마지막으로 로컬에 반영한 서버 잔액 기준선 — '증가분'만 더하기 위함. */
    private const val LAST_SYNCED_SERVER_KEY = "withu.genQuota.lastSyncedServer.v1"

    private val prefs get() = SharedAppState.prefs()

    /** 오늘 yyyymmdd. 자정 넘으면 값이 바뀌어 일일 카운트 자동 리셋. */
    private fun todayStamp(calendar: Calendar = Calendar.getInstance()): Int =
        calendar.get(Calendar.YEAR) * 10000 +
            (calendar.get(Calendar.MONTH) + 1) * 100 +
            calendar.get(Calendar.DAY_OF_MONTH)

    /** 오늘 일일 한도에서 사용한 횟수. */
    fun usedToday(): Int =
        computeUsedToday(prefs.getInt(DATE_KEY, 0), prefs.getInt(COUNT_KEY, 0), todayStamp())

    /** 충전 크레딧 잔액 (만료 없음). */
    fun credits(): Int = prefs.getInt(CREDITS_KEY, 0)

    /** 지금 더 만들 수 있는 총 횟수. DEBUG 빌드는 9999 (무제한 테스트 — 출시 빌드는 실제 한도). */
    fun remainingToday(): Int =
        if (BuildConfig.DEBUG) 9999
        else max(0, freeDailyLimit - usedToday()) + credits()

    fun canGenerate(n: Int = 1): Boolean = remainingToday() >= n

    /**
     * n 회 사용 기록 — 오늘 한도부터 차감, 부족분은 크레딧에서.
     * DEBUG 는 no-op (차감 안 함). 반드시 생성 '성공 시에만' 호출할 것.
     */
    fun record(n: Int = 1) {
        if (BuildConfig.DEBUG) return
        if (n <= 0) return
        val today = todayStamp()
        val usedBase = computeUsedToday(prefs.getInt(DATE_KEY, 0), prefs.getInt(COUNT_KEY, 0), today)
        val result = computeRecord(n, usedBase, credits(), freeDailyLimit)
        val e = prefs.edit()
        e.putInt(DATE_KEY, today)
        e.putInt(COUNT_KEY, result.newUsed)
        if (result.creditsChanged) e.putInt(CREDITS_KEY, result.newCredits)
        e.apply()
    }

    /**
     * 퀄리티별 캔디 비용 (프레임 1장 기준) — high 6 · medium 3 · 기본(low) 1.
     * 움직이는 캐릭터는 frame0·frame1 을 각각 생성·과금하므로 호출 측이 프레임마다 record —
     * 여기서 2배 계산하지 말 것.
     */
    fun cost(forQuality: String): Int = when (forQuality) {
        "high" -> 6
        "medium" -> 3
        else -> 1
    }

    /** 캔디 팩 구매/코드 적용 시 크레딧 적립. */
    fun addCredits(n: Int) {
        if (n <= 0) return
        prefs.edit().putInt(CREDITS_KEY, credits() + n).apply()
    }

    /**
     * 서버 잔액(entitlement.credits)을 로컬에 반영 — delta > 0 만 가산.
     * 기준선이 없으면(최초 동기화) 서버 잔액이 더 클 때만 한 번 끌어올림. 기준선은 항상 갱신.
     */
    fun syncCreditsUp(serverCredits: Int) {
        val lastSynced = if (prefs.contains(LAST_SYNCED_SERVER_KEY)) {
            prefs.getInt(LAST_SYNCED_SERVER_KEY, 0)
        } else null
        val newCredits = computeSyncedCredits(credits(), lastSynced, serverCredits)
        prefs.edit()
            .putInt(CREDITS_KEY, newCredits)
            .putInt(LAST_SYNCED_SERVER_KEY, serverCredits)
            .apply()
    }

    /**
     * 로그아웃 시 서버 잔액 기준선 리셋 — 같은 기기 다른 계정의 delta 왜곡 방지.
     * 로컬 credits 는 유지 (미로그인 구매분 보호).
     */
    fun resetServerBaseline() {
        prefs.edit().remove(LAST_SYNCED_SERVER_KEY).apply()
    }

    /**
     * 화면 표시용 보유 캔디 = 로컬 잔액 그대로.
     * canGenerate 가 로컬 credits 만 보므로 배지도 같은 값 — 과대표시 방지.
     */
    fun displayedCandy(): Int = credits()

    // MARK: - 순수 코어 (단위 테스트 대상 — DEBUG 분기/SharedPreferences 와 분리)

    internal fun computeUsedToday(storedStamp: Int, storedCount: Int, todayStamp: Int): Int =
        if (storedStamp == todayStamp) storedCount else 0

    internal data class RecordResult(
        val newUsed: Int,
        val newCredits: Int,
        val creditsChanged: Boolean,
    )

    /** 오늘 무료 한도부터 차감, 부족분은 크레딧에서 (음수 방지 clamp). */
    internal fun computeRecord(n: Int, usedBase: Int, credits: Int, dailyAllowance: Int): RecordResult {
        val dailyLeft = max(0, dailyAllowance - usedBase)
        val fromDaily = min(n, dailyLeft)
        val fromCredits = n - fromDaily
        return RecordResult(
            newUsed = usedBase + fromDaily,
            newCredits = if (fromCredits > 0) max(0, credits - fromCredits) else credits,
            creditsChanged = fromCredits > 0,
        )
    }

    /**
     * syncCreditsUp 코어 — lastSynced 가 null(최초)이면 서버가 더 클 때만 끌어올리고,
     * 이후엔 delta > 0 만 가산 (delta < 0 은 무시 — 기준선만 갱신).
     */
    internal fun computeSyncedCredits(localCredits: Int, lastSynced: Int?, serverCredits: Int): Int {
        if (lastSynced == null) {
            return if (serverCredits > localCredits) serverCredits else localCredits
        }
        val delta = serverCredits - lastSynced
        return if (delta > 0) localCredits + delta else localCredits
    }
}
